tests/common/mod.rs

```diff
@@ -23,52 +35,102 @@ fn num_diff_bits(s1: &[u8], s2: &[u8]) -> usize {
+    let r2 = f(&mut v2).unwrap();
 
     // Between 3.5 and 4.5 bits per byte should differ. Probability of failure:
     // ~ 2^(-94) = 2 * CDF[BinomialDistribution[8000, 0.5], 3500]
-    let d = num_diff_bits(&v1, &v2);
+    let d = num_diff_bits(r1, r2);
     assert!(d > 3500);
     assert!(d < 4500);
 }
 
-// Tests the quality of calling getrandom repeatedly on small buffers
+#[cfg(not(feature = "custom"))]
 #[test]
+fn test_large() {
+    test_diff_large(0u8, wrapped_getrandom);
+}
+
 #[cfg(not(feature = "custom"))]
-fn test_small() {
+#[test]
+fn test_large_uninit() {
+    test_diff_large(MaybeUninit::uninit(), getrandom_uninit_impl);
+}
+
+// Tests the quality of calling getrandom repeatedly on small buffers
+
+#[cfg(not(feature = "custom"))]
+fn test_diff_small<T: Copy>(initial: T, f: impl Fn(&mut [T]) -> Result<&mut [u8], Error>) {
     // For each buffer size, get at least 256 bytes and check that between
     // 3 and 5 bits per byte differ. Probability of failure:
     // ~ 2^(-91) = 64 * 2 * CDF[BinomialDistribution[8*256, 0.5], 3*256]
     for size in 1..=64 {
         let mut num_bytes = 0;
         let mut diff_bits = 0;
         while num_bytes < 256 {
-            let mut s1 = vec![0u8; size];
-            getrandom_impl(&mut s1).unwrap();
-            let mut s2 = vec![0u8; size];
-            getrandom_impl(&mut s2).unwrap();
+            let mut s1 = vec![initial; size];
+            let r1 = f(&mut s1).unwrap();
+            let mut s2 = vec![initial; size];
+            let r2 = f(&mut s2).unwrap();
 
             num_bytes += size;
-            diff_bits += num_diff_bits(&s1, &s2);
+            diff_bits += num_diff_bits(r1, r2);
         }
         assert!(diff_bits > 3 * num_bytes);
         assert!(diff_bits < 5 * num_bytes);
     }
 }
 
+#[cfg(not(feature = "custom"))]
+#[test]
+fn test_small() {
+    test_diff_small(0u8, wrapped_getrandom);
+}
+
+#[cfg(not(feature = "custom"))]
+#[test]
+fn test_small_unnit() {
+    test_diff_small(MaybeUninit::uninit(), getrandom_uninit_impl);
+}
+
 #[test]
 fn test_huge() {
     let mut huge = [0u8; 100_000];
     getrandom_impl(&mut huge).unwrap();
 }
 
+#[test]
+fn test_huge_uninit() {
+    let mut huge = [MaybeUninit::uninit(); 100_000];
+    getrandom_uninit_impl(&mut huge).unwrap();
+    check_initialized(&huge);
+}
+
+#[allow(unused_variables)]
+fn check_initialized(buf: &[MaybeUninit<u8>]) {
+    #[cfg(feature = "unstable-sanitize")]
+    {
+        // XXX: `#![feature(cfg_sanitize)]` doesn't enable the feature gate correctly.
+        // #[cfg(sanitize = "memory")]
+        {
+            use core::ffi::c_void;
+            extern "C" {
+                // void __msan_check_mem_is_initialized(const volatile void *x, size_t size);
+                fn __msan_check_mem_is_initialized(x: *const c_void, size: usize);
+            }
+            unsafe {
+                __msan_check_mem_is_initialized(buf.as_ptr().cast::<c_void>(), buf.len());
+            }
+        }
+    }
+}
```

Comment:
Instead of this complexity, can we just do something in the huge tests which reads the `huge` buffer. Then MSAN will fail if we've messed up, but we don't need to have any `cfg`-sepecific code in our tests.

Reply:
Yes, we could maybe do the same thing we do for the large tests? I didn't want to try understanding what the large test is doing, so this was easier :) Also I was curious about whether `__msan_check_mem_is_initialized` would work.